### PR TITLE
fix: Allow to pass any object with .toJSON method to JSON field

### LINKED ISSUE #20192: `Error: Unknown value type` on nested create
### Bug description

After upgrading to prisma v5, we are now seeing the following error when executing a nested create query. We have downgraded to 4.16.2 as a workaround. `DEBUG=prisma*` did not yield any more relevant logs for the error. Will attempt to make a small repro.

```
error - Error: Unknown value type
    at pe (/.../node_modules/@prisma/client/runtime/library.js:26:706)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8408)
    at ia (/.../node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/.../node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/.../node_modules/@prisma/client/runtime/library.js:115:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400)
    at ia (/.../node_modules/@prisma/client/runtime/libra15:8562)
    at na (/.../node_modules/@prisma/client/runtime/library.js:115:8400) {
  clientVersion: '5.0.0',
  ```

### How to reproduce

1. Run nested create query

### Expected behavior

Nested create query succeeds.

### Prisma information

```ts
datasource db {
  provider = "postgres"
  url      = env("TEST_POSTGRES_URI")
}

generator client {
  provider = "prisma-client-js"
  output   = "../node_modules/.prisma/client"
}

model User {
  uuid           String           @id @default(uuid())
  teamUuid       String           @default(uuid())
  userName       String
  UserTemplateV2 UserTemplateV2[]
}

model UserTemplateV2 {
  uuid   String @id @default(uuid())
  userId String
  from   String
  jsonField Json
  user   User   @relation(fields: [userId], references: [uuid])
}
```

```ts
await prisma.user.create({
  select: { uuid: true, teamUuid: true },
  data: {
    userName: name.trim(),
    UserTemplateV2: {
      create: [
        {
          from: `${user.firstName} ${user.lastName}`,
          jsonField: { href: new URL("http://www.example.com") }
        },
        {
          from: `${user.firstName} ${user.lastName}`,
          jsonField: { href: new URL("http://www.example.com") }
        },
      ],
    },
  },
});
```

### Environment & setup

- OS: Alpine Linux
- Database: PostgreSQL 14
- Node.js version: v16.19.0

### Prisma Version

```
5.0.0
```

### PR DESCRIPTION
**Pull request by SevInf**
That used to work in GraphQL protocol and JS. Technically, it's an
"accidental" feature and it does not work on TS level. After some
internal discussions, we've decided to make it a proper feature (and
also fix TS part of it).

One more thing this PR does: for non-serializable values, it will show
rich error message with context, rather internal one, thrown by
`assertNever`. Since query arguments come from untrusted user input,
relying on a fixed type for them was incorrect to begin with.
Fix #20192

### CODE REVIEW

File: packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts [hunk 1]
```diff
@@ -227,11 +225,24 @@ function serializeArgumentsValue(
...
-  assertNever(jsValue, 'Unknown value type')
+  context.throwValidationError({
+    kind: 'InvalidArgumentValue',
+    selectionPath: context.getSelectionPath(),
+    argumentPath: context.getArgumentPath(),
+    argument: {
+      name: context.getArgumentName(),
+      typeNames: [],
+    },
+    underlyingError: `Don't knows how to serialize ${Object.prototype.toString.call(jsValue)} value`,
```
Comment (Jolg42) [inline packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts:244]:
```suggestion
    underlyingError: `Does't know how to serialize ${Object.prototype.toString.call(jsValue)} value. Serialize the object to JSON or implement a ".toJSON()" method on it.`,
```

File: packages/client/tests/functional/naming-conflict/_builtInNames.ts [hunk 2]
```diff
@@ -1,59 +1,60 @@
...
-  'LogEvent',
-  'PrismaAction',
-  'MiddlewareParams',
-  'Middleware',
-  'SortOrder',
-  'TransactionClient',
-  'IntFilter',
-  'NestedIntFilter',
-  'BatchPayload',
+  'Query',
```
Comment (Jolg42) [inline packages/client/tests/functional/naming-conflict/_builtInNames.ts:2]: Why the comments for everything except Query here?

File: packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts [hunk 3]
```diff
@@ -227,11 +225,24 @@ function serializeArgumentsValue(
...
-  assertNever(jsValue, 'Unknown value type')
+  context.throwValidationError({
+    kind: 'InvalidArgumentValue',
+    selectionPath: context.getSelectionPath(),
+    argumentPath: context.getArgumentPath(),
+    argument: {
+      name: context.getArgumentName(),
+      typeNames: [],
+    },
+    underlyingError: `Does't know how to serialize ${Object.prototype.toString.call(jsValue)} value. Serialize the object to JSON or implement a ".toJSON()" method on it.`,
```
Comment (Jolg42) [inline packages/client/src/runtime/core/jsonProtocol/serializeJsonQuery.ts:244]:
```suggestion
    underlyingError: `Doesn't know how to serialize ${Object.prototype.toString.call(jsValue)} value. Serialize the object to JSON or implement a ".toJSON()" method on it.`,
```
sorry for my typo

**Review by Jolg42 — APPROVED** [review]

**Review by aqrln — APPROVED** [review]

File: packages/client/src/runtime/core/errorRendering/applyValidationError.ts [hunk 4]
```diff
@@ -321,15 +321,18 @@ function applyInvalidArgumentValueError(error: InvalidArgumentValueError, args:
...
       parts.push(`: ${error.underlyingError}`)
     }
-    parts.push(`. Expected ${expected}.`)
+    parts.push('.')
+    if (error.argument.typeNames.length > 0) {
+      const expected = joinWithPreposition(
+        'or',
+        error.argument.typeNames.map((type) => colors.green(type)),
+      )
+      parts.push(` Expected ${expected}.`)
```
[the remark this answers is not in the document]
Comment (SevInf) [reply packages/client/src/runtime/core/errorRendering/applyValidationError.ts:334]: In older tests, extra `.` came from `underlyingError`: engine-side errors don't include dot, so we add it to the message here. We could check if it's already there and only add it when necessary, but ultimately, it is easier to to just remove it from client-side  `underlyingError`:

**Pull request merged**